Add example of disclosure navigation with top-level links

smhigley opened the pull request:
Adds a new disclosure menu example page with top-level links in addition to disclosures, based on the 10/27 APG meeting. Related to #89. Also updates the disclosure nav js file to use a class pattern based on the updated code style guide.

[Preview Link](https://raw.githack.com/w3c/aria-practices/disclosure-nav2/examples/disclosure/disclosure-navigation-hybrid.html)

### Review checklist

* [x] [Editorial review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#editorial_review) by @mcking65 
* [x] [Functional review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#functional_review) by @carmacleod 
* [x] [Visual design review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#design_review) by @a11ydoer 
* [x] [Accessibility review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#accessibility_review) by @shirsha 
* [x] [Code review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#code_review) by @spectranaut 
* [x] [Test review](https://github.com/w3c/aria-practices/wiki/Pull-Request-Review-Process#test_review) by @spectranaut


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/w3c/aria-practices/pull/1614.html" title="Last updated on Mar 3, 2021, 7:18 PM UTC (00c1af9)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/aria-practices/1614/7f56b9e...00c1af9.html" title="Last updated on Mar 3, 2021, 7:18 PM UTC (00c1af9)">Diff</a>

carmacleod commented:
Looks good! Keyboard works well, i.e. tab, arrows, home, end, space, enter, escape - and various combinations of tab and arrows - all work intuitively. Will try screen readers later.

Couple of notes:
- would be cool to have a [CodePen button](https://github.com/w3c/aria-practices/wiki/How-to-add-the-Open-In-Codepen-button-to-an-example-page)
- the dropdowns show, then hide, on page load - would be great if they were initially hidden

carmacleod commented on examples/disclosure/disclosure-navigation-hybrid.html:
```diff
@@ -0,0 +1,376 @@
+      <li>Since this set of disclosure buttons controls visibility of navigation links for the mythical university web site, the list that contains them is wrapped in a navigation landmark named <q>Mythical University</q>.</li>
+      <li>
+        The semantics of the list structure communicates the hierarchy of the navigation system to assistive technology users.
+        The top-level set of three buttons is in a list of three items, and the set of links controlled by each button is contained within a list nested inside the item with the controlling button.
```
This point may need a bit of wordsmithing, since there are now 3 links and 3 buttons, making 6 items in the top-level list.

carmacleod replied:
Never mind! Just looked at the DOM, and I see that it's coded as 3 list items, each with link and disclosure button children. Makes sense! Resolving this convo.

mcking65 commented:
@smhigley  asks:

> @mcking65 -- was there ever an issue filed for this one?

No, we didn't create an issue for it. We can just use this PR. I'll add labels here.

spectranaut reviewed (approved):
This is an approving code review, looks good!

Can you add the [open in codepen button](https://github.com/w3c/aria-practices/wiki/How-to-add-the-Open-In-Codepen-button-to-an-example-page)?

For tests: so much of the code and functionality is the same between this example and "Disclosure for Navigation Menus" is the same. You can certainly just copy over the relevant tests.

mcking65 commented:
@smhigley , looking awesome!

I have two questions.

First, The buttons are labeled like "Show About section". The word "section" doesn't quite feel right; makes me think it is going to show a section of content. I'm wondering if it should be something like "show links for About section" or "Show links to About pages"

Second, as part of a quest for consistency among related examples, would you be OK with making the sample content look and feel the way Jon has it set up in the tree example and upcoming new version of the menubar example? See #1558. On those pages, he kind of made a mini web site inside the example. It has a header, footer, and the main content region is labeled by the site name and H1. One of his objectives was to put a couple focusable elements inside the main content region. Note that he used role region, not main. There are comments about that in the source and in the documentation tables.

spectranaut commented:
Hi @smhigley! Thanks for adding the codepen button. If you notice, the links don't work in a very friendly way in Codepen, it takes you back to the the APG webpage when you click on them. @carmacleod and I looked into ways to make the example work on both the APG webpage and in codepen on this similar example you wrote: https://github.com/w3c/aria-practices/pull/1651#issuecomment-739050311

What do you think of preventing the default behavior of the onclick button so that if will work in the codepen example?

spectranaut reviewed (approved):
Just looked over your tests, @smhigley , this is an approving test review!

smhigley commented:
@spectranaut Thanks for the review! I'll add the preventDefault to the example links, and push it up together with a test fix, as soon as I figure out what's making them fail 😅

nschonni commented:
I think `npm run reference-tables` should fix the example failure for you

carmacleod commented on examples/disclosure/disclosure-navigation-hybrid.html:
```diff
@@ -0,0 +1,378 @@
+        <ul id="exTest" class="disclosure-nav">
+          <li>
+            <a href="#mythical-page-content" class="main-link">About</a>
+            <button type="button" aria-expanded="true" aria-controls="id_about_menu" aria-label="Show About section"> </button>
+            <ul id="id_about_menu">
```
Can we start all of the buttons off in the collapsed state so that there's no initial flash when the ul is hidden?
```suggestion
            <button type="button" aria-expanded="false" aria-controls="id_about_menu" aria-label="Show About section"> </button>
            <ul id="id_about_menu" style="display: none;">
```

carmacleod commented on examples/disclosure/disclosure-navigation-hybrid.html:
```diff
@@ -0,0 +1,378 @@
+          </li>
+          <li>
+            <a href="#mythical-page-content" class="main-link">Admissions</a>
+            <button type="button" aria-expanded="true" aria-controls="id_admissions_menu" aria-label="Show Admissions section"> </button>
+            <ul id="id_admissions_menu">
```
Prevent initial flash.
```suggestion
            <button type="button" aria-expanded="false" aria-controls="id_admissions_menu" aria-label="Show Admissions section"> </button>
            <ul id="id_admissions_menu" style="display: none;">
```

carmacleod commented on examples/disclosure/disclosure-navigation-hybrid.html:
```diff
@@ -0,0 +1,378 @@
+          </li>
+          <li>
+            <a href="#mythical-page-content" class="main-link">Academics</a>
+            <button type="button" aria-expanded="true" aria-controls="id_academics_menu" aria-label="Show Academics section"> </button>
+            <ul id="id_academics_menu">
```
Prevent initial flash.
```suggestion
            <button type="button" aria-expanded="false" aria-controls="id_academics_menu" aria-label="Show Academics section"> </button>
            <ul id="id_academics_menu" style="display: none;">
```

smhigley replied:
I thought we did this so that if JS weren't enabled or didn't load, it would still work. Is that not a concern we're targeting?

carmacleod reviewed (commented):
Not sure if this is the prettiest way to prevent the initial show-hide flash, but it does work. :)

a11ydoer commented:
@smhigley 
The [preview link](https://raw.githack.com/w3c/aria-practices/disclosure-nav2/examples/disclosure/disclosure-navigation-hybrid.html) above does not show changes such as label of "Show About section", which @mcking65 commented on. I just want to make sure that I review the most recent example. Thanks for your consideratoin.

a11ydoer commented:
I checked the current example with VoiceOver and it works well. Keyboard navigation also works as documented.

a11ydoer commented:
I could not find the same visual implementation from popular sites yet. However, the closest visual design as well as its coding to current menu example can be found at WAI tutorial, [Flyout menu, use button as toogle.(https://www.w3.org/WAI/tutorials/menus/flyout/#use-button-as-toggle)] The screen shot was attached. 
![Uploading flyout menu design with arrow and expanded submenu items.png…]()

a11ydoer commented:
I am waiting for the most updated preview link before I finish accessiblity review.

smhigley commented:
@a11ydoer the preview link is up to date. I don't see any comment from Matt on the button labels -- what was the feedback?

I'm also not sure I understand your comment on the WAI tutorial. Were you suggesting we add a link to it from the example page? I'd be happy to add one if so; just let me know where you want it :).

a11ydoer commented:
@smhigley thanks for the updated preview link. 
Matt's comment is [here](https://github.com/w3c/aria-practices/pull/1614#issuecomment-745585062)

You can ignore my wai example comment. it is simple justification method for visual review by me. I am trying to make sense for APG design with other design example.

carmacleod commented:
@smhigley Thanks for hiding the dropdowns on load!

When the user clicks on a link, should focus go to the mythical-page-content div?

jscholes commented:
@smhigley Glad to see an example for this pattern in the APG!  IMHO, the labels for the disclosure buttons are very verbose for screen reader users, and all start with the same three syllables ("Show links to") rather than the most significant information being placed upfront.

As a related note, the verb "show" is probably unnecessary, because the use of `aria-expanded` already indicates that this is a disclosure and it isn't updated when the buttons become expanded anyway.

Not sure on the best wording, but I'd definitely like to see the name of the section come first.  Maybe just, "About sub-menu", "About section", "About pages", or similar?

a11ydoer commented:
@smhigley 

Here are two suggestions from APG meeting on Feb 9, 2021

1. aria-label change: upon @jscholes suggestion, the "show links to about pages" becomes "links to about pages"
2. Title changes to "Example disclosure navigation menu with top level links"

smhigley commented:
@jscholes / @a11ydoer the disclosure button label was @mcking65's suggestion, so I want to make sure he gets the chance to review and approve the change. I'm happy to change it to anything :).

Another possibility might be something simple like "Admissions submenu" -- I know "submenu" and "menu" means something specific in ARIA, but for general labeling I think it might fit well here.

@a11ydoer I'll update that title too :)

jscholes commented:
@smhigley 

> Another possibility might be something simple like "Admissions submenu" -- I know "submenu" and "menu" means something specific in ARIA, but for general labeling I think it might fit well here.

FWIW, this wording (sub-menu) is what I recommend to clients implementing such a pattern.  I've come across quite a few cases of this pattern being implemented mostly correctly in the wild, but with really verbose messaging for the buttons.  This example list only has three disclosures, but it's not unusual to have six or more.

mcking65 commented:
@jscholes, @smhigley ,

In APG [section 5.3.3 Composing Effective and User-friendly Accessible Names](https://w3c.github.io/aria-practices/#naming_effectively), one point of guidance we provide is:

>Do NOT include a WAI-ARIA role name in the accessible name. For example, do not include the word “ button ”  in the name of a button, the word “ image ”  in the name of an image, or the word “ navigation ”  in the name of a navigation region.

In this case, putting the word "menu" in a button label would communicate that the button opens a menu. It would be indistinguishable from an actual menu button for most users.

If the label were "About submenu", then most screen readers will read that as "About submenu button collapsed". This is concerning to me if one of our long term goals is to build consistency in user experiences because this button does not open a menu.

While the label would be using the word menu in the colloquial sense, and one could argue that is the sense in which most users interpret most of what they hear, it still weakens our ability to create greater harmony between expected behaviors and application behavior.

I am also concerned that writing guidance that justifies using some role names in some labels but not others  would introduce a new, unnecessary source of confusion for authors. This is especially important because we tell authors that roles communicate expected behaviors. In the menu examples, we tell authors not to use the menu role unless they are going to provide those behaviors.

I see the appeal of the label "About submenu". It is short and sweet. It communicates the meaning of the down arrow icon pretty well if we were to ignore everything I just wrote above. There is a lot of meaning packed into the visual design that it is difficult to capture with just a couple of words.

On the other hand, if we want labels that are clear, we want a label that tells the screen reader user what this button does. The button shows a list of links to other pages; those pages are related to the page that is linked just before the button; that list will appear in the reading order just after the button.

The expanded/collapsed semantics are intended to capture the show/hide behavior. And, we do want users to expect that the content that is shown or hidden immediately follows in the reading order. Those are the fundamentals of the disclosure pattern.

So, the main concepts we need to communicate are:
1. What content will be exposed by activating the button.
2. What is the relationship between that content and surrounding content.

The name "Links to about pages" is intended to capture both of these objectives. If we shorten it to "About Pages", I am concerned that we don't sufficiently communicate the type of content that is exposed for most users. It is a pretty subtle judgment, though, I will admit.

As I am writing this, though, another option that I think may be even better comes to mind:

>More About pages

The word "more" can do a good job of capturing the "sub" nature of the relationship. And , since the element that precedes is a link, it also communicates the idea that what is exposed is links.

I think if this labeling were replicated by people reading the example page, it would work in nearly any scenario.
Our accessible naming guidance

jscholes commented:
@mcking65 

> More About pages

Works for me!

mcking65 commented:
I have a few questions as I review the content.

1. Are we doing anything with borders to support focus indication in high contrast mode that should be documented under accessibility features? We document features to support high contrast mode in other examples.

2. Is this statement in accessibility features still applicable?

>Implementing this Esc behavior is necessary to meet the WCAG 2.1 1.4.13: Content on Hover or Focus criterion. 

It doesn't seem that we trigger anything to appear on hover or focus, so it seems it would not apply.

3. How is the expanded and collapsed state visually indicated? I'd like to make the following statement a bit more specific for all readers and more understandable for blind readers:

>The visual indicator of the show/hide state is created using CSS ::after  pseudo element border styles so the caret is reliably rendered in high contrast mode of operating systems and browsers.

mcking65 commented:
@smhigley 

Could you change the name of the disclosure buttons to "More {name} pages", e.g., "More About pages"?

I also noticed focus moves to the region instead of the heading when a link is activated. In the tree and menubar examples, we are moving it to the heading. Would you be OK with making this the same?

mcking65 commented:
@smhigley 
I finished the title change. However, after you rebase and we have the latest menubar and tree nav examples in this branch, we should fix the names of the links that reference the disclosure navigation menu examples on those pages.

mcking65 commented:
Should we put the advisement class on the note about the menu role? And, I wonder if we should also add clarification similar to what we offer on the navigation menubar ppage. Something like:

><strong>Note:</strong> Although this example uses the word “menu” in the colloquial sense to refer to a set of navigation links, it does not use the WAI-ARIA menu role.
>That is because The <code>menu</code> and <code>menubar</code>  roles require complex functionality, such as <a href="../../#kbd_general_within">composite wwidget focus management</a> and first-character navigation,  that is unnecessary for typical site navigation.

carmacleod reviewed (approved):
+1 This is an approving Functional review.

Nice work, @smhigley !

carmacleod commented:
@mcking65 
> Is this statement in accessibility features still applicable?
>> Implementing this Esc behavior is necessary to meet the WCAG 2.1 1.4.13: Content on Hover or Focus criterion.
> It doesn't seem that we trigger anything to appear on hover or focus, so it seems it would not apply.

I agree that this sentence can be deleted, but the rest of the sentences in that point should stay.

> How is the expanded and collapsed state visually indicated?

In normal mode, the button gets a blue background color and the "V" within the button goes from black to white. In high contrast mode, there is no background color change (which is fine), and the "V" is the color of text, and can be easily seen.

a11ydoer commented:
APG Group will work on focus styling documentation in the future.

carmacleod commented:
@mcking65 
> Should we put the advisement class on the note about the menu role? And, I wonder if we should also add clarification similar to what we offer on the navigation menubar page. 

I think this makes sense. We might even move the note up so that it follows the green "Important note", for consistency with other advisement blocks.

Note that the clarification sentence as written has a couple of typos: capitalized The in mid-sentence, and wwidget with 2 w's. Otherwise, the words seem fine to me.

smhigley commented:
@mcking65 I added the advisement text you recommended to the example page (with the typo fixes @carmacleod caught).

I also removed the manual focus handling in favor of just resetting the link targets for Codepen (the other approach I mentioned on the call). After thinking about it, I like this approach better since it only relies on built-in browser behavior, and reduces the chance of anyone thinking that javascript focus handling is needed in this pattern.

carmacleod commented:
@smhigley 
> I added the advisement text 

Great - thanks! Please note that Matt's suggested sentence contained this link: [composite widget focus management](https://github.com/w3c/#kbd_general_within)

> I also removed the manual focus handling in favor of just resetting the link targets

Nice fix for the CodePen problem! We should probably do that in the other nav examples, too (where possible).

I think focus should go to the fake page when the user clicks on a link. :)

shirsha commented:
**NVDA/Chrome**: When any link in the disclosure button ("More {name} pages") is selected, it is narrated as "unknown".
**JAWS/ Chrome**: The links in the disclosure is selected, it is announced with some extra verbose. Ex: Facts -  "n fact sample cont"; Signup - "md signup sample content"

https://user-images.githubusercontent.com/7432373/108931089-ae3f4800-760c-11eb-955b-61187e6669ee.mp4

https://user-images.githubusercontent.com/7432373/108931105-b39c9280-760c-11eb-9576-f31993239d61.mp4

smhigley commented:
@shirsha I can't reproduce with JAWS 2020, NVDA 2020, and Chrome 88. The link behavior is entirely handled by the browser though, so I don't think it would be related to the code in this example.

shirsha commented:
@smhigley  I used JAWS 2021 , NVDA 2020.3 , Chrome  88.0.4324.190

a11ydoer reviewed (approved):
- aria-label,  more [name] pages was confirmed through the review.
- emphasis on how this example is different from aria menu looks good.
- focus styling,  having a few pixels out of border line will be researched later. 

Thanks for all your work @smhigley !

a11ydoer commented:
@shirsha 
Thanks for the recorded screen reader videos! These are very useful. 

I noticed that when the screen reader announced "signup sample content" the focus was already in content section, not in the "sign up" sublink.

mcking65 commented:
I have finished making editorial changes for consistency. I have also fixed the title of and links to the other disclosure navigation menu example.

There is one part of the text that I'm still not satisfied is sufficiently clear. It is related to high contrast mode documentation. It is:

>To ensure the carets that indicate the presence and state of the disclosures are reliably rendered when operating system or browser high contrast settings are enabled, visual indication of the show/hide state is created using CSS ::after  pseudo element border styles.

It really doesn't yet explain why this is choice of coding technique matters. Compare to what we have written in the navigation menu bar for similar HCM support.

>To ensure the arrow icons used to indicate the expanded or collapsed state have sufficient contrast with the background when high contrast settings invert colors, the CSS currentColor  value for the fill  and stroke  properties of the SVG polygon  element is used to synchronize the color with text content. If specific colors are used to specify the fill  and stroke  properties, these colors will remain the same in high contrast mode, which could lead to insufficient contrast between the icon and the background or even make the icon invisible if its color matches the high contrast mode background.

This text from the navigation menubar explains the feature, why it matters, and why the technique for achieving it was chosen.

Could someone suggest wording that achieves these same objectives for the HCM documentation for this example?

mcking65 commented:
This is all ready to merge except that I would still like some assistance improving the high contrast documentation in accessibility features as described in my previous comment. Could someone help me understand the rationale for choice of the after pseudo border style and what distinguishes that technique from others when it comes to HCM support?

mcking65 commented:
Per discussion in March 2, 2021 meeting, I removed:

> To ensure the carets that indicate the presence and state of the disclosures are reliably rendered when operating system or browser high contrast settings are enabled, visual indication of the show/hide state is created using CSS ::after  pseudo element border styles.

Conducting a final read and will merge.

mcking65 reviewed (approved):
This approving editorial review.

Love this addition! Thank you @smhigley!
